Add access to Search2 index using search/record api

xmorave2 opened the pull request:
TODO:

- [x] Do we need to add SearchApiRecordFields yaml config for search2 too?
- [x] Update wiki page when merged

demiankatz commented:
Thanks, @xmorave2, this looks like a good start. I'll be interested to hear @EreMaijala's thoughts on this, but he is currently out of the office for a few weeks, so we won't get feedback right away.

Regarding SearchApiRecordFields.yaml, I think it is likely there will be use cases where the two services will need to be differentiated. You should be able to take advantage of the `@parent_yaml` inheritance mechanism (see [this code](https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/Config/YamlReader.php#L148)) to set up a default Search2ApiRecordFields.yaml that simply inherits from the standard one -- that way we don't have more configurations to maintain, but there's an easy place to make customizations when needed.

xmorave2 commented:
Thanks @demiankatz, I added the second yaml file

This is not critical for us now, so I have no problem with waiting to @EreMaijala's review

EreMaijala commented on module/VuFindApi/config/module.config.php:
```diff
@@ -52,6 +55,28 @@
+                'type' => 'Laminas\Router\Http\Literal',
+                'verb' => 'get,post,options',
+                'options' => [
+                    'route'    => '/api/v1/search2',
```
How about using something like '/api/v1/index2/search' and '/api/v1/index2/record' as the routes? This would allow the endpoints to look to the same for any other indexes and operations too, and I feel that it would be better to avoid exposing the "search2" name.

xmorave2 replied:
Does make sense, I changed it in follow-up

xmorave2 commented:
Thanks @EreMaijala

demiankatz commented:
Thanks, @EreMaijala for the review, and @xmorave2 for the ongoing work. A couple of thoughts/issues:

1.) When I go to the http://my-server/vufind/api URL now, I see both sets of API endpoints ("/record" / "/search" and "/index2/record" / "/index2/search") but the descriptions are identical. Maybe we should add a new variable to swagger.phtml so we can differentiate the descriptions and better explain which index is being accessed by each API endpoint. Otherwise, I think this is potentially confusing.

2.) I configured my Search2.ini, and the /index2/search endpoint seems to be working as expected. However, when I try to look up a record from the search results using the /index2/record endpoint, I always seem to get a 400 response:

<pre>
{
  "status": "ERROR",
  "statusMessage": "Error loading record"
}
</pre>

Do I have something set up incorrectly, or are you seeing similar results?

xmorave2 commented:
Thank for comments @demiankatz 

ad 1) I added the index identification (primary/secondary) to all routes description

ad 2) You are right, getting record was not working as expected, i added searchClassId parameter to $recordLoader->load(), it should be working now

demiankatz commented:
Thanks, @xmorave2, this is working for me now. @EreMaijala, any objections to merging this at this point?

One minor outstanding issue is that I think it may be beneficial to actually put the index labels into the controllers rather than determining them conditionally in swagger.phtml. I plan on adding another endpoint for the website index after this is merged, and once we have three possible descriptions, putting conditions in the template will start to make a lot less sense... but I think making that change can be done as part of the website API work, so I'm happy to merge this as-is for now if there are no bigger concerns to hold us up.
